### ISSUE TITLE
If no DSS equipped, landable bodies are described as not landable

### ISSUE THREAD

**Issue by richardbuckle** [EDCD/EDDI#107]
## Steps:

1. Equip a Discovery Scanner but not a Detailed Surface Scanner.
2. Go to a system with a landable body that you haven't scanned yet.
3. Scan the body.

## Expected:
EDDI either says that the body is landable or stays silent on the matter.

## Observed:
EDDI falsely says the body is not landable.

## Investigation
In BodyScannedEvent.cs, EddiEvents.BodyScannedEvent.BodyScannedEvent()

```
VARIABLES.Add("landable", "True if the body is landable (only available if DSS equipped)");
```

The problem is that if you scan a landable body without a DSS, it defaults `landable` to false rather than leaving it as `null` or some other value representing "unknown". This causes EDDI to incorrectly say that the body is not suitable for landing.

Here is the scan JSON:
```
{
  "timestamp": "2017-09-24T22:50:13Z",
  "event": "Scan",
  "BodyName": "HIP 19026 B 2 e",
  "DistanceFromArrivalLS": 30124.728516,
  "PlanetClass": "Rocky body",
  "MassEM": 0.00299,
  "Radius": 1010799.6875,
  "SurfaceGravity": 1.166228,
  "SemiMajorAxis": 3236784640.0,
  "Eccentricity": 0.000198,
  "OrbitalInclination": -0.766237,
  "Periapsis": 148.165283,
  "OrbitalPeriod": 1808504.875,
  "RotationPeriod": 1808529.25,
  "AxialTilt": -0.08085
}
```
and here is the speech output (using my "Succint Explorer" personality):
```
Scan of Body B 2 e finished. This is a low-gravity Rocky body that is unsuitable for landing.
```

**Comment by richardbuckle** [comment]
Maybe wait to see if ED2.4 provides more info in the scan event. Landability is visible in the UI so it should be in the scan event too I would argue.

**Comment by Syrrianyl** [comment]
The problem is that with these data, it is impossible to say if a body is landable or not. At what I see, landable have "SurfacePressure":0.000000, and if the scanner is not "detailled" there is no "SurfacePressure" data.
Can the missing "SurfacePressure" data be used to say that the landable is "unknown"?

**Comment by richardbuckle** [comment]
If you have a DSS then along with the "SurfacePressure" variable you get an explicit "landable" Boolean variable. If you have no DSS then both are missing even though the system map still shows whether the body is landable. So the presence or absence of the "SurfacePressure" variable doesn't give us any new info.

Let's see what 2.4 gives us before starting work.

**Comment by Hoodathunk** [comment]
This has been a problem for some time because there's been no reliable way to determine the presence of a DSS.

I think the addition of the 'module' events and the updating of the hardpoint/compartment data in the ship object present an opportunity to automatically detect particular modules and provide a boolean ('hasDSS', for example) to filter/not use unreliable data

**Comment by richardbuckle** [comment]
In this case the data in the event JSON is accurate, it's just incomplete -- it's our default assumption of "landable = false" which isn't necessarily correct. It's really a JSON parsing issue. The fact that it occurs when you don't have a DSS is kind of incidental.

**Comment by Tkael** [comment]
I think it's basically just the Cottle script. rather than just returning landable if true, it should return landable if true, not landable if false, and shouldn't comment on the subject if it's null.

**Comment by Tkael** [comment]
I've also been exploring some options for better integrating BodyDetails with local data (in an experimental branch) to fill holes in the data.

**Comment by richardbuckle** [comment]
Ah you could be right. I'll need to do another test and verify that `landable` is indeed null. I had the impression that it was being incorrectly set to false but I might have been wrong.

**Comment by Hoodathunk** [comment]
TryGetValue() will set the value to the default if the key is not found... which in the of Booleans is false.

This is your culprit...

`        private static bool? getOptionalBool(IDictionary<string, object> data, string key)
        {
            object val;
            data.TryGetValue(key, out val);
            return getOptionalBool(key, val);
        }
`
Changing to...

`        private static bool? getOptionalBool(IDictionary<string, object> data, string key)
        {
            object val;
            if (data.TryGetValue(key, out val))
                return getOptionalBool(key, val);
            else
                return null;
        }
`

would give you the 'null' you want

**Comment by richardbuckle** [comment]
Cheers for that. I'll give it a test this evening.

**Comment by Tkael** [comment]
I agree that having variables in the ship object corresponding to certain modules would be a nice upgrade. Not sure if it should be part of this issue or separated?

**Comment by richardbuckle** [comment]
Separated I think, and probably post-2.4.0?

**Comment by richardbuckle** [comment]
Incidentally this second implementation of `getOptionalBool()` is nonsense:

```c#
        private static bool? getOptionalBool(string key, object val)
        {
            return (bool?)val;
        }
```
Fortunately it's not used anywhere.

**Comment by richardbuckle** [comment]
Pull request #110

**Comment by richardbuckle** [comment]
Fixed for 2.4-b4
